.github/workflows/50_go_build-one-arch-one-os-one-path.yml

```diff
@@ -0,0 +1,141 @@
+      go-mod-cache-restore-or-update:
+        required: false
+        type: string
+        default: 'restore'
+      goarch:
+        required: false
+        type: string
+        default: ''   # default is to take it from ${{ runner.arch }}
+      goos:
+        required: false
+        type: string
+        default: ''   # default is to take it from ${{ runner.os }}
+      pack-to-docker-image:
+        required: false
+        type: boolean
+        default: false
+      base-sha:
+        required: false
+        type: string
+        default: 'anysha'
+
+
+jobs:
+  build:
+    name: "${{ inputs.goos }}/${{ inputs.goarch }}"
+    runs-on: [ self-hosted, small ]
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-go@v5
+        id: setup-go
+        with:
+          cache: false
+          go-version-file: ${{ inputs.go-mod-path }}
+      - id: go-cache-paths
+        name: Set Cache Paths
+        run: |
+          echo "go-build=$(go env GOCACHE)" >> $GITHUB_OUTPUT
+          echo "go-mod-cache=$(go env GOMODCACHE)" >> $GITHUB_OUTPUT
+      - name: Set GOARCH and GOOS
+        id: goarch-goos
+        run: |
+          GOARCH_INPUT="${{ inputs.goarch }}"
+          GOOS_INPUT="${{ inputs.goos }}"
+          declare -A runner_to_go_os=( ["Linux"]="linux" ["Windows"]="windows" ["macOS"]="darwin" )
+          declare -A runner_to_go_arch=( ["X86"]="386" ["X64"]="amd64" ["ARM"]="arm" ["ARM64"]="arm64" )
+          ARCH_FROM_RUNNER="${runner_to_go_arch[${{ runner.arch }}]}"
+          OS_FROM_RUNNER="${runner_to_go_os[${{ runner.os }}]}"
+          echo "goarch=${GOARCH_INPUT:-$ARCH_FROM_RUNNER}" >> "$GITHUB_OUTPUT"
+          echo "goos=${GOOS_INPUT:-$OS_FROM_RUNNER}" >> "$GITHUB_OUTPUT"
+      - id: set-cache-keys
+        name: Set keys needed for caching
+        run: |
+          OS="${{ steps.goarch-goos.outputs.goos }}"
+          ARCH="${{ steps.goarch-goos.outputs.goarch }}"
+          GO_VER="${{ steps.setup-go.outputs.go-version }}"
+          GO_SUM_HASH="${{ hashFiles('**/go.sum') }}"
+          GH_REPO="${{ github.event.repository.full_name }}"
+          GH_REPO="${GH_REPO//\//_}"  # '//\//_' wat?
+                                      # here how it works:
+                                      # ${var//pattern/replace},
+                                      # and pattern is '/', so escape it: '\/'.
+                                      # and replace is just '_', easy.
+          GIT_SHA="${{ inputs.base-sha }}"
+          CMD_PATH="${{ inputs.cmd-path }}"
+          CMD_PATH="${CMD_PATH//\//_}"  # well, now you know what '//\//_' means and how it works, right? :)
+
+          echo "go-mod-cache-key=go-mod-cache-${OS}-${ARCH}-go-${GO_VER}-${GO_SUM_HASH}" >> "$GITHUB_OUTPUT"
+          for run_target in cmd test; do
+            echo "go-build-cache-key-${run_target}=go-build-cache-${OS}-${ARCH}-go-${GO_VER}-gh-${GH_REPO}-git-${GIT_SHA}-${run_target}-${CMD_PATH}" >> "$GITHUB_OUTPUT"
+          done
```

Comment:
the test key isn't used so far? I'd not add code with the intention of using it in the future, I'd add that once we're there.

Reply:
As I said, I'm hesitant to add dead code, so I'd go for dropping the loop for now

Reply:
I though how long will it take for me to add tests, and it was mostly copy-paste from build, so I added test job, so now no deadcode